bytes/methods.mbt

```diff
@@ -13,21 +13,107 @@
-    for j in 0..<pattern_len {
-      guard target.unsafe_get(i + j) == pattern.unsafe_get(j) else { break }
-    } else {
-      return Some(i)
+  guard pattern_len > 0 else { return Some(0) }
+  guard target_len >= pattern_len else { return None }
+  if pattern_len == 1 {
+    // Optimized single character search
+    let pattern_byte = pattern.unsafe_get(0)
+    for i in 0..<target_len {
+      if target.unsafe_get(i) == pattern_byte {
+        return Some(i)
+      }
+    }
+    return None
+  }
+
+  // Build KMP failure function (border array)
+  let border = build_border_array(pattern)
+
+  // Build Boyer-Moore bad character skip table
+  let skip_table = build_skip_table(pattern)
+  fjs_search(target, pattern, border, skip_table)
+}
+
+///|
+/// Build KMP border array (failure function)
+fn build_border_array(pattern : View) -> Array[Int] {
+  let pattern_len = pattern.length()
+  let border = Array::make(pattern_len, 0)
+  let mut j = 0
+  for i in 1..<pattern_len {
+    while j > 0 && pattern.unsafe_get(i) != pattern.unsafe_get(j) {
+      j = border[j - 1]
     }
-  } else {
-    None
+    if pattern.unsafe_get(i) == pattern.unsafe_get(j) {
+      j += 1
+    }
+    border[i] = j
+  }
+  border
+}
+
+///|
+/// Build Boyer-Moore bad character skip table
+fn build_skip_table(pattern : View) -> Array[Int] {
+  let pattern_len = pattern.length()
+  let skip_table = Array::make(256, pattern_len)
+
+  // Fill skip table with distances from rightmost occurrence
+  for i in 0..<(pattern_len - 1) {
+    let byte_val = pattern.unsafe_get(i).to_int()
+    skip_table[byte_val] = pattern_len - 1 - i
+  }
+  skip_table
+}
+
+///|
+/// FJS hybrid search algorithm
+fn fjs_search(
+  target : View,
+  pattern : View,
+  border : Array[Int],
+  skip_table : Array[Int],
+) -> Int? {
+  let target_len = target.length()
+  let pattern_len = pattern.length()
+  let last_pattern_byte = pattern.unsafe_get(pattern_len - 1)
+  let mut i = 0 // position in target
+  while i <= target_len - pattern_len {
+    // Boyer-Moore style: check last character first
+    let target_byte = target.unsafe_get(i + pattern_len - 1)
+    if target_byte != last_pattern_byte {
+      // Bad character skip
+      let skip = skip_table[target_byte.to_int()]
+      i += skip
+      continue
+    }
+
+    // Last character matches, now use KMP-style matching from left
+    let mut j = 0 // position in pattern
+    let mut k = i // current position in target
+    while j < pattern_len &&
+          k < target_len &&
+          target.unsafe_get(k) == pattern.unsafe_get(j) {
+      j += 1
+      k += 1
+    }
+    if j == pattern_len {
+      return Some(i) // Found match
+    }
+
+    // Mismatch occurred, use KMP shift or Boyer-Moore shift, whichever is larger
+    let kmp_shift = if j == 0 { 1 } else { j - border[j - 1] }
+    let bm_shift = 1 // Minimum shift for Boyer-Moore when last char matched
+    i += if kmp_shift > bm_shift { kmp_shift } else { bm_shift }
```

Comment:
I don't think this is correct. If `j > 0` after KMP match failure, we should retry KMP matching starting from `j <- border[j - 1]`. But the code here will repeat the comparison of `pattern[0:border[j-1]]`. This may lead to `O(n * m)` complexity in the worst case, for example when matching `a*ba` against `a*`.

If `j > 0`, we should just retry the KMP match with updated `j`, until `j == 0`.